### PR TITLE
C interface for Meshfree Operators

### PR DESCRIPTION
**Pull request by sslattery**
This PR updates the C interface in several ways:

1. Adds an interface for creating, applying, and destroying DTK map operators.
2. Adds a preliminary implementation of this interface using the `NearestNeighbor` operator. 
    - The intention is to generalize the interface for construction by using a factory and expanding the implementation to use the other DTK operators in a future PR.
3. Adds an additional `MemorySpace` concept to the C interface.
    - DTK maps operate in `ExecutionSpaces` and are therefore constructed from them - this is where the computation happens.
    - DTK user application interfaces operate in `MemorySpaces` and are therefore constructed from them - this is where user access to DTK memory happens.
4. Implements and tests using different memory spaces for the source and target along with different execution spaces for the map.

### CODE REVIEW

**Comment by sslattery** [comment]
Preliminary demonstration implementation is done - this is not final but is a proof of concept. Import notes:
- New `MapFactory` package currently does not use ETI. It will not build unless ETI is off because of undefined references to the ETI symbols from the rest of the library.
- Turning off ETI took my build time from 30 minutes to 16. @dalg24 maybe we want to experiment with this on the CI.

File: packages/Interface/cmake/Dependencies.cmake [hunk 1]
```diff
@@ -4,4 +4,7 @@ TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
   Kokkos
   Teuchos
   Tpetra
+
+  LIB_REQUIRED_TPLS
+  MPI
```
Comment (dalg24) [inline packages/Interface/cmake/Dependencies.cmake]: Sure but then let's reflect that in the documentation.

We have been getting back and forth on this, whether MPI should be optional or required.  If the `Interface` package requires it, and all other packages providing solution transfer operator (`Meshfree` and `Discretization`) have a required dependency on the `Interface` package, then DTK pretty much requires MPI.

File: packages/Interface/src/DTK_C_API.h [hunk 2]
```diff
@@ -149,6 +222,17 @@ extern void DTK_finalize();
...
+ *
+ * DTK error codes.
+ */
+typedef enum {
+    DTK_SUCCESS = 0,
+    DTK_INVALID_HANDLE = -1,
+    DTK_UNINITIALIZED = -2,
+    DTK_UNKNOWN = -99
+} DTK_Error;
+
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: How did it compile before?

File: packages/Interface/src/DTK_C_API.h [hunk 3]
```diff
@@ -149,6 +222,17 @@ extern void DTK_finalize();
 
 /**@}*/
 
+/** \brief Get DTK error message.
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: Comment is duplicated after the declaration of `DTK_Error`.
Make sure we look at the documentation generated with Sphinx before merging.

File: packages/MapFactory/src/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#
+# A) Package-specific configuration options
+#
+
+TRIBITS_CONFIGURE_FILE(${PACKAGE_NAME}_config.h)
```
Comment (dalg24) [inline packages/MapFactory/src/CMakeLists.txt]: File is empty.  Remove `DataTransferKitMapFactory_config.h.in` and comment this out.
I'd like to move away from every package having its own `config.h`.

File: packages/MapFactory/src/DTK_C_API_Map.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <DTK_C_API.h>
+#include "DTK_C_API_Map.hpp"
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.cpp]: Why did you decide on double quote over angle bracket on this include?

File: packages/Interface/src/DTK_C_API.h [hunk 6]
```diff
@@ -94,16 +98,85 @@ extern DTK_UserApplicationHandle DTK_create( DTK_ExecutionSpace space );
...
+ *  Must be created using DTK_createMap() to be a valid handle.
+ *
+ *  The handle essentially hides C++ implementation details from the user.
+ *
+ *  <!--
+ *  Use incomplete types to differentiate between handles.
+ *  We never define the incomplete structs.
+ *  -->
+ */
+typedef struct _DTK_MapHandle *DTK_MapHandle;
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: Do we prefer "Map" over "Operator"?

File: packages/MapFactory/src/DTK_C_API_Map.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                             DTK_UserApplicationHandle target )
+{
+    if ( !DTK_is_initialized() )
+    {
+        errno = DTK_UNINITIALIZED;
+        return nullptr;
+    }
+
+    // For demonstration purposes just use the nearest neighbor map.
+    auto handle = reinterpret_cast<DTK_MapHandle>(
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.cpp]: I am not sure how I feel about this `reinterpret_cast`.
Should `createMap` return type be `DTK_MapHandle` if its only purpose is to support the C interface?

File: packages/MapFactory/src/DTK_C_API_Map.cpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+//---------------------------------------------------------------------------//
+void DTK_destroyMap( DTK_MapHandle handle )
+{
+    if ( DataTransferKit::valid_map_handles.count( handle ) )
+    {
+        auto dtk = reinterpret_cast<DataTransferKit::DTK_Map*>( handle );
+        delete dtk;
+        DataTransferKit::valid_map_handles.erase( handle );
+    }
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.cpp]: No else-clause?
If you attempt to destroy a handle that is not listed as valid, should the error code be `DTK_SUCCESS`?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+/*!
+ * \file
+ * \brief C adapter to DTK Maps.
+ */
+#ifndef DTK_C_API_MAP_HPP
+#define DTK_C_API_MAP_HPP
+
+#include <memory>
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: This include is ~~not placed properly.~~ duplicated below.

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+#include <DataTransferKit_config.hpp>
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: We need to have a discussion about these `_config.hpp` includes.

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                 DTK_UserApplicationHandle source,
+                 DTK_UserApplicationHandle target )
+        : _source( reinterpret_cast<DTK_Registry*>(source)->_registry )
+        , _target( reinterpret_cast<DTK_Registry*>(target)->_registry )
+    {
+        // Create a Teuchos comm from the MPI comm.
+        auto teuchos_comm = Teuchos::rcp( new Teuchos::MpiComm<int>(comm) );
+
+        // FOR NOW JUST CREATE A NEAREST NEIGHBOR OPERATOR FOR DEMONSTRATION
+        // PURPOSES. THIS WILL BE REPLACED BY A PROPER FACTORY.
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: You planning on adding arguments to the constructor to specify which map to build?
Have you given much thoughts on what these might be?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        for ( int t = 0; t < num_tgt; ++t )
+            target_field.dofs(t,0) = target_field_copy(t);
+
+        // Push the data to the target.
+        _target.pushField( target_field_name, target_field );
+    }
+
+    UserApplication<double,Device> _source;
+    UserApplication<double,Device> _target;
+    std::shared_ptr<NearestNeighborOperator<Device> > _map;
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: Why a `shared_ptr`?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+        // Copy to a compatible layout. Nearest neighbor operator needs a
+        // compatible layout and only 1 dimension
+        int num_src = source_field.dofs.extent(0);
+        Kokkos::View<double*,Device>
+            source_field_copy("source_field_copy", num_src );
+        for ( int s = 0; s < num_src; ++s )
+        {
+            source_field_copy(s) = source_field.dofs(s,0);
+        }
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: Can't you just use a `Kokkos::subview` instead.

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+// Create a map.
+DTK_Map* createMap( DTK_ExecutionSpace space,
+                    MPI_Comm comm,
+                    DTK_UserApplicationHandle source,
+                    DTK_UserApplicationHandle target )
+{
+    switch ( space )
+    {
+        case DTK_SERIAL:
+#ifdef KOKKOS_HAVE_SERIAL
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: This has been deprecated in favor of `KOKKOS_ENABLE_SERIAL`.

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        {
+            return new DTK_MapImpl<Serial>( comm, source, target );
+        }
+#else
+        {
+            DTK_INSIST( false );
+            return nullptr;
+        }
+#endif
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: No `break;`?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+{
+    virtual ~DTK_Map() = default;
+
+    virtual void apply( const std::string& source_field_name,
+                        const std::string& target_field_name ) = 0;
+};
+
+//---------------------------------------------------------------------------//
+template<class Device>
+struct DTK_MapImpl : public DTK_Map
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: Not sure I get why you introduced inheritance here.  Can you explain?

File: packages/Interface/src/DTK_C_API.h (hunk 6 above)
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: I think "Map" is a bit more intuitive for those without an extensive linear algebra background. It simply "maps" the data from one code to another. Although I guess "TransferOperator" is also equally intuitive as long as the word "transfer" is included.

File: packages/MapFactory/src/DTK_C_API_Map.hpp (hunk 11 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.hpp]: For now I'm hoping that a string of parameters is sufficient to build the maps. A map should know which user data functions it needs to populate its data structures and if those functions aren't implemented a user should get clear error messages.

File: packages/Interface/src/DTK_ParallelTraits.hpp [hunk 17]
```diff
@@ -25,99 +25,34 @@
 namespace DataTransferKit
```
Comment (dalg24) [inline packages/Interface/src/DTK_ParallelTraits.hpp]: This header make little sense to me.  It is only pulling few Kokkos types into the DTK namespace.

File: packages/Interface/test/tstC_API.cpp [hunk 18]
```diff
@@ -65,34 +65,24 @@ int check_registry( const char *name, DTK_UserApplicationHandle handle )
 
     switch ( dtk->_space )
     {
-    case DTK_SERIAL:
-#ifdef KOKKOS_HAVE_SERIAL
+    case DTK_HOST_SPACE:
+#if defined( KOKKOS_HAVE_SERIAL ) || defined( KOKKOS_HAVE_OPENMP )
```
Comment (dalg24) [inline packages/Interface/test/tstC_API.cpp]: `KOKKOS_ENABLE_`

File: packages/Interface/test/tstKokkosView.cpp [hunk 19]
```diff
@@ -210,14 +210,60 @@ TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( View, basic_reduce_kernel, Scalar,
...
+
+    // Create a view in the execution space.
+    using ViewType = Kokkos::View<Scalar *, MemorySpace>;
+    const int size = 1000;
+    ViewType data( "data", size );
+    ViewType result( "result", size );
+
+    // Populate the view in the execution space.
+    Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, size ),
+                          FillFunctor<ViewType>( data ) );
```
Comment (dalg24) [inline packages/Interface/test/tstKokkosView.cpp]: `FillFunctor` is not a good name. I was going to tell you to use `Kokkos::deep_copy()` and then Iooked at the implementation and I realized you meant `iota( data )`

File: packages/Interface/test/tstKokkosView.cpp [hunk 20]
```diff
@@ -210,14 +210,60 @@ TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( View, basic_reduce_kernel, Scalar,
...
     for ( int i = 0; i < size; ++i )
     {
         test_sum += host_data( i );
     }
     TEST_EQUALITY( test_sum, sum );
 }
 
+//---------------------------------------------------------------------------//
+// Test for creating a view and running a basic scan kernel
+TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( View, basic_scan_kernel, Scalar, DeviceType )
```
Comment (dalg24) [inline packages/Interface/test/tstKokkosView.cpp]: What is the point of this unit test?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    UserApplication<double, SourceMemSpace> _source;
+    UserApplication<double, TargetMemSpace> _target;
+    std::unique_ptr<NearestNeighborOperator<map_device_type>> _map;
+};
+
+//---------------------------------------------------------------------------//
+// Execution space validation.
+bool validateExecutionSpace( DTK_ExecutionSpace space )
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: `validExecutionSpace`?

File: packages/MapFactory/src/DTK_C_API_Map.hpp [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+    switch ( map_space )
+    {
+    case DTK_SERIAL:
+        switch ( src_space )
+        {
+        case DTK_HOST_SPACE:
+            switch ( tgt_space )
+            {
+            case DTK_HOST_SPACE:
+                map = new DTK_MapImpl<Serial, HostSpace, HostSpace>(
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.hpp]: You need to protect these with preprocessor directives `#if defined( KOKKOS_ENABLE_SERIAL )`

File: packages/Interface/src/DTK_C_API.h [hunk 23]
```diff
@@ -94,16 +101,85 @@ extern DTK_UserApplicationHandle DTK_create( DTK_ExecutionSpace space );
...
-extern bool DTK_is_valid( DTK_UserApplicationHandle handle );
+extern bool DTK_isValidUserApplication( DTK_UserApplicationHandle handle );
 
-/** \brief Destroy a DTK handle.
+/** \brief Destroy a DTK handle to a user application.
  *
  *  \param[in,out] handle User application handle.
  */
-extern void DTK_destroy( DTK_UserApplicationHandle handle );
+extern void DTK_destroyUserApplication( DTK_UserApplicationHandle handle );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: yeah sorry we needed another destroy function for the map and I wanted them named consistently

File: packages/Interface/src/DTK_ParallelTraits.hpp [hunk 24]
```diff
@@ -25,99 +25,34 @@
...
-// Serial
-#if defined( KOKKOS_HAVE_SERIAL )
-using Serial = Kokkos::Serial;
-#endif
-
-// OpenMP
-#if defined( KOKKOS_HAVE_OPENMP )
-using OpenMP = Kokkos::OpenMP;
+// Host memory.
+#if defined( KOKKOS_HAVE_SERIAL ) || defined( KOKKOS_HAVE_OPENMP )
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: will fix

File: packages/Interface/src/DTK_ParallelTraits.hpp (hunk 17 above)
Comment (sslattery) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: I agree. my next interface refactor will likely remove this and clean up some.

File: packages/Interface/test/tstC_API.cpp (hunk 18 above)
Comment (sslattery) [reply packages/Interface/test/tstC_API.cpp]: will fix

File: packages/Interface/test/tstKokkosView.cpp (hunk 19 above)
Comment (sslattery) [reply packages/Interface/test/tstKokkosView.cpp]: going to get rid of this test so N/A

File: packages/Interface/test/tstKokkosView.cpp (hunk 20 above)
Comment (sslattery) [reply packages/Interface/test/tstKokkosView.cpp]: this is from when I was trying to figure out if the cuda 9 issues came from parallel_scan. This doesn't belong in this PR so I am removing the test.

File: packages/MapFactory/src/DTK_C_API_Map.hpp (hunk 21 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.hpp]: sure

File: packages/MapFactory/src/DTK_C_API_Map.hpp (hunk 22 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.hpp]: good catch

**Comment by sslattery** [comment]
There is an OpenMP variant of the C interface test failing on 2 of the 4 builds. I will look into this.

**Comment by sslattery** [comment]
@dalg24 this is cleaned up if you want to take another look

File: packages/Discretization/src/DTK_FE.hpp [hunk 25]
```diff
@@ -34,6 +34,8 @@
 #include <Intrepid2_HGRAD_WEDGE_C1_FEM.hpp>
 #include <Intrepid2_HGRAD_WEDGE_C2_FEM.hpp>
 
+#include <memory>
```
Comment (dalg24) [inline packages/Discretization/src/DTK_FE.hpp]: Did you mean to add this include directive?

File: packages/Interface/cmake/Dependencies.cmake (hunk 1 above)
Comment (dalg24) [reply packages/Interface/cmake/Dependencies.cmake]: You haven't addressed this comment.  I'd be fine resolving it in a later PR but I want us to open an issue to track it.

File: packages/Interface/src/DTK_ParallelTraits.hpp (hunk 17 above)
Comment (dalg24) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: Why postponing?

File: packages/MapFactory/src/DTK_C_API_Map.cpp (hunk 7 above)
Comment (dalg24) [reply packages/MapFactory/src/DTK_C_API_Map.cpp]: Ping.

File: packages/MapFactory/src/DTK_C_API_Map.cpp [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+    return DataTransferKit::valid_map_handles.count( handle );
+}
+
+//---------------------------------------------------------------------------//
+void DTK_applyMap( DTK_MapHandle handle, const char *source_field,
+                   const char *target_field )
+{
+    if ( !DTK_isValidMap( handle ) )
+    {
+        errno = DTK_INVALID_HANDLE;
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.cpp]: Please trigger this in your test.

File: packages/MapFactory/src/DTK_C_API_Map.cpp [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+    if ( DataTransferKit::valid_map_handles.count( handle ) )
+    {
+        auto dtk = reinterpret_cast<DataTransferKit::DTK_Map *>( handle );
+        delete dtk;
+        DataTransferKit::valid_map_handles.erase( handle );
+        errno = DTK_SUCCESS;
+    }
+    else
+    {
+        errno = DTK_INVALID_HANDLE;
```
Comment (dalg24) [inline packages/MapFactory/src/DTK_C_API_Map.cpp]: Test for this too.

File: packages/Interface/cmake/Dependencies.cmake (hunk 1 above)
Comment (sslattery) [reply packages/Interface/cmake/Dependencies.cmake]: I think it should be required for DTK as a whole. See #418

File: packages/Interface/src/DTK_C_API.h (hunk 2 above)
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: no idea

File: packages/MapFactory/src/DTK_C_API_Map.cpp (hunk 7 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.cpp]: This is needed to hide the C++ class pointer we actually use under the opaque handle type. Not sure you can do this with a static cast or any other type of cast.

File: packages/Interface/src/DTK_ParallelTraits.hpp (hunk 17 above)
Comment (sslattery) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: I figured this PR was big enough with the clear purpose of introducing the map interface. The next PR would make more infrastructure changes to the C interface.

File: packages/Discretization/src/DTK_FE.hpp (hunk 25 above)
Comment (sslattery) [reply packages/Discretization/src/DTK_FE.hpp]: yeah not sure why that is there.

Comment (sslattery) [reply packages/Discretization/src/DTK_FE.hpp]: oh yeah he was using `std::unique_ptr`

Comment (sslattery) [reply packages/Discretization/src/DTK_FE.hpp]: https://github.com/ORNL-CEES/DataTransferKit/blob/36ee35f08b39856b077604accdb57570ead8b335/packages/Discretization/src/DTK_FE.hpp#L247

Comment (dalg24) [reply packages/Discretization/src/DTK_FE.hpp]: OK.  Sorry I should have realized you have a separate commit for that one.  Good!

File: packages/Interface/src/DTK_ParallelTraits.hpp (hunk 17 above)
Comment (dalg24) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: Ok.  Would you open an issue to capture this?  Get rid of `DTK_ParallelTraits.hpp` and probably remove "imports" of Cuda/OpenMP/Serial into DTK's namespace.

File: packages/MapFactory/src/DTK_C_API_Map.cpp (hunk 26 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.cpp]: done

File: packages/MapFactory/src/DTK_C_API_Map.cpp (hunk 27 above)
Comment (sslattery) [reply packages/MapFactory/src/DTK_C_API_Map.cpp]: done

**Comment by sslattery** [comment]
also I will clean history before we merge.

File: packages/Interface/src/DTK_ParallelTraits.hpp (hunk 17 above)
Comment (sslattery) [reply packages/Interface/src/DTK_ParallelTraits.hpp]: See #420

**Pull request merged**